mods/bones/init.lua

```diff
@@ -181,24 +199,70 @@ end
 local player_inventory_lists = { "main", "craft" }
 bones.player_inventory_lists = player_inventory_lists
 
-local function is_all_empty(player_inv)
+--functions registered this way won't becalled if bones_mode is keep
+function bones.register_dead_player_inv_management(func)
+	table.insert(dead_player_callbacks, func)
+end
+
+local function transfer_stack_to_bones(stk,current_dead_player)
+	-- check if it's possible to place bones, if not find space near player
+	if ( current_dead_player.bones_mode == "bones" ) and
+			( current_dead_player.bones_pos == nil ) then
+		current_dead_player.bones_pos = current_dead_player.player_pos
+		local air
+		if may_replace(current_dead_player.bones_pos, current_dead_player.player) then
+			air = current_dead_player.bones_pos
+		else
+			air = minetest.find_node_near(current_dead_player.bones_pos, 1, {"air"})
+		end
+
+		if air and not minetest.is_protected(air, current_dead_player.player_name) then
+			current_dead_player.bones_pos = air
+			local param2 = minetest.dir_to_facedir(current_dead_player.player:get_look_dir())
+			minetest.set_node(current_dead_player.bones_pos, {name = "bones:bones", param2 = param2})
+			local meta = minetest.get_meta(current_dead_player.bones_pos)
+			current_dead_player.bones_inv = meta:get_inventory()
+			--make it so big that anything reasonable will for sure fit inside
+			current_dead_player.bones_inv:set_size("main", theoretical_max_slots)
+		else
+			current_dead_player.bones_mode = "drop"
+			current_dead_player.bones_pos = nil
+		end
+	end
+
+	if ( current_dead_player.bones_mode == "bones" ) and
+			( current_dead_player.bones_inv:room_for_item("main", stk) ) then
+		current_dead_player.bones_inv:add_item("main", stk)
+	else
+		drop(current_dead_player.player_pos, stk)
+		current_dead_player.dropped=true
+	end
+end
+
+local function player_dies_transfer_inventory(player,transfer_stack)
+	local player_inv = player:get_inventory()
 	for _, list_name in ipairs(player_inventory_lists) do
-		if not player_inv:is_empty(list_name) then
-			return false
+		for i = 1, player_inv:get_size(list_name) do
+			local stack = player_inv:get_stack(list_name, i)
+			transfer_stack(stack)
 		end
+		player_inv:set_list(list_name, {})
 	end
-	return true
 end
 
+bones.register_dead_player_inv_management(player_dies_transfer_inventory)
+
 minetest.register_on_dieplayer(function(player)
+	local pos = vector.round(player:get_pos())
 	local bones_mode = minetest.settings:get("bones_mode") or "bones"
 	if bones_mode ~= "bones" and bones_mode ~= "drop" and bones_mode ~= "keep" then
 		bones_mode = "bones"
 	end
+	local player_name = player:get_player_name()
+	local current_dead_player={player=player, player_name=player_name, bones_inv=nil, bones_pos=nil,
```

Comment:
Stuffing all variables into this table isn't clean at all. You could easily get rid of this by just inlining all the code here as it was before, and inserting the appropriate callback calls.

If you want to split this into functions, do it properly, with well-defined functions and proper arguments rather than opaque tables.